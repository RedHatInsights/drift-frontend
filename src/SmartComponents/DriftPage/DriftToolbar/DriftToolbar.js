import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { DropdownItem, PaginationVariant, Toolbar, ToolbarGroup, ToolbarItem,
    ToolbarContent, ToolbarFilter } from '@patternfly/react-core';

import FilterDropDown from '../FilterDropDown/FilterDropDown';
import SearchBar from '../SearchBar/SearchBar';
import ActionKebab from '../ActionKebab/ActionKebab';
import AddSystemButton from '../AddSystemButton/AddSystemButton';
import ExportCSVButton from '../../ExportCSVButton/ExportCSVButton';
import { TablePagination } from '../../Pagination/Pagination';

export class DriftToolbar extends Component {
    constructor(props) {
        super(props);

        this.state = {
            actionKebabItems: [
                <DropdownItem
                    key="remove-systems"
                    component="button"
                    data-ouia-component-id='clear-all-comparisons-dropdown-item'
                    onClick={ this.clearComparison }>Clear all comparisons</DropdownItem>
            ],
            dropdownItems: [
                <DropdownItem
                    key='export-to-CSV'
                    component='button'
                    data-ouia-component-id='export-to-csv-dropdown-item-comparison'
                    onClick={ () => this.props.exportToCSV() }
                >
                    Export to CSV
                </DropdownItem>
            ],
            isEmpty: true,
            dropdownOpen: false
        };
    }

    setFactFilterChips = () => {
        const { activeFactFilters, factFilter } = this.props;
        let factFilterChips = [ ...activeFactFilters ];

        if (factFilter.length && !activeFactFilters.includes(factFilter)) {
            factFilterChips.push(factFilter);
        }

        return factFilterChips;
    }

    async clearAllFactChips() {
        const { activeFactFilters, filterByFact, handleFactFilter } = this.props;

        await activeFactFilters.forEach(function (filter) {
            handleFactFilter(filter);
        });

        filterByFact('');
    }

    setStateChips = (stateFilters) => {
        let stateChips = [];

        stateFilters.forEach(function(filter) {
            if (filter.selected) {
                stateChips.push(filter.display);
            }
        });

        return stateChips;
    }

    clearAllStateChips = async () => {
        const { addStateFilter, stateFilters } = this.props;

        stateFilters.forEach(function(stateFilter) {
            stateFilter.selected = true;
            addStateFilter(stateFilter);
        });
    }

    removeChip = async (type = '', id = '') => {
        const { activeFactFilters, addStateFilter, clearAllFactFilters, filterByFact, handleFactFilter, setHistory, stateFilters } = this.props;

        if (type) {
            if (type === 'State') {
                if (id === '') {
                    this.clearAllStateChips();
                } else {
                    stateFilters.forEach(async function(stateFilter) {
                        if (stateFilter.display === id) {
                            await addStateFilter(stateFilter);
                        }
                    });
                }
            } else {
                if (id === '') {
                    await clearAllFactFilters();
                } else if (activeFactFilters.includes(id)) {
                    await handleFactFilter(id);
                } else {
                    await filterByFact('');
                }
            }
        } else {
            await this.clearAllStateChips();
            await this.clearAllFactChips();
        }

        setHistory();
    }

    setIsEmpty = (isEmpty) => {
        this.setState({ isEmpty });
    }

    onToggle = () => {
        const { dropdownOpen } = this.state;

        this.setState({
            dropdownOpen: !dropdownOpen
        });
    }

    clearFilters = async () => {
        const { clearComparisonFilters, setHistory } = this.props;

        await clearComparisonFilters();
        setHistory();
    }

    clearComparison = async () => {
        const { clearComparison, clearSelectedBaselines, setHistory, setIsFirstReference, updateReferenceId } = this.props;

        await clearComparison();
        await clearSelectedBaselines('CHECKBOX');
        await setIsFirstReference(true);
        await updateReferenceId();
        setHistory();

    }

    render() {
        const { activeFactFilters, factFilter, filterByFact, handleFactFilter, loading,
            page, perPage, setHistory, stateFilters, totalFacts, updatePagination } = this.props;
        const { actionKebabItems, dropdownItems, dropdownOpen } = this.state;

        return (
            <React.Fragment>
                <Toolbar className="drift-toolbar" clearAllFilters={ this.removeChip }>
                    <ToolbarContent>
                        <ToolbarGroup variant='filter-group'>
                            <ToolbarFilter
                                chips={ this.setFactFilterChips() }
                                deleteChip={ this.removeChip }
                                deleteChipGroup={ this.removeChip }
                                categoryName="Fact name"
                            >
                                <SearchBar
                                    factFilter={ factFilter }
                                    activeFactFilters={ activeFactFilters }
                                    handleFactFilter={ handleFactFilter }
                                    filterByFact={ filterByFact }
                                    setHistory={ setHistory }
                                />
                            </ToolbarFilter>
                            <ToolbarFilter
                                chips={ this.setStateChips(stateFilters) }
                                deleteChip={ this.removeChip }
                                deleteChipGroup={ this.removeChip }
                                categoryName="State"
                            >
                                <FilterDropDown setHistory={ setHistory } />
                            </ToolbarFilter>
                        </ToolbarGroup>
                        <ToolbarGroup variant='button-group'>
                            <ToolbarItem>
                                <AddSystemButton loading={ loading } />
                            </ToolbarItem>
                        </ToolbarGroup>
                        <ToolbarGroup variant='icon-button-group'>
                            <ToolbarItem>
                                <ExportCSVButton
                                    dropdownItems={ dropdownItems }
                                    ouiaId='export-dropdown-comparison'
                                    isOpen={ dropdownOpen }
                                    onToggle={ this.onToggle }
                                />
                            </ToolbarItem>
                            <ToolbarItem>
                                <ActionKebab
                                    ouiaId='clear-comparison-dropdown'
                                    dropdownItems={ actionKebabItems } />
                            </ToolbarItem>
                        </ToolbarGroup>
                        <ToolbarItem variant='pagination' align={{ default: 'alignRight' }}>
                            <TablePagination
                                page={ page }
                                perPage={ perPage }
                                total={ totalFacts }
                                isCompact={ true }
                                updatePagination={ updatePagination }
                                widgetId='drift-pagination-top'
                                ouiaId='comparison-pagination-top'
                                variant={ PaginationVariant.top }
                            />
                        </ToolbarItem>
                    </ToolbarContent>
                </Toolbar>
            </React.Fragment>
        );
    }
}

DriftToolbar.propTypes = {
    loading: PropTypes.bool,
    history: PropTypes.object,
    page: PropTypes.number,
    perPage: PropTypes.number,
    totalFacts: PropTypes.number,
    updatePagination: PropTypes.func,
    clearComparisonFilters: PropTypes.func,
    clearComparison: PropTypes.func,
    exportToCSV: PropTypes.func,
    clearSelectedBaselines: PropTypes.func,
    setIsFirstReference: PropTypes.func,
    updateReferenceId: PropTypes.func,
    factFilter: PropTypes.string,
    filterByFact: PropTypes.func,
    stateFilters: PropTypes.array,
    addStateFilter: PropTypes.func,
    activeFactFilters: PropTypes.array,
    handleFactFilter: PropTypes.func,
    clearAllFactFilters: PropTypes.func,
    setHistory: PropTypes.func
};

export default DriftToolbar;
