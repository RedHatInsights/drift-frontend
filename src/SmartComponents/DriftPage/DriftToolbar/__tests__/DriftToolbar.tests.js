import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import DriftToolbar from '../DriftToolbar';
import stateFilters from '../../../modules/__tests__/state-filter.fixtures';

describe('DriftToolbar', () => {
    let props;

    beforeEach(() => {
        props = {
            loading: false,
            page: 1,
            perPage: 50,
            totalFacts: 30,
            factFilter: '',
            activeFactFilters: [],
            clearSelectedBaselines: jest.fn(),
            clearComparison: jest.fn(),
            clearComparisonFilters: jest.fn(),
            updateReferenceId: jest.fn(),
            updatePagination: jest.fn(),
            setIsFirstReference: jest.fn(),
            history: { push: jest.fn() },
            stateFilters: stateFilters.allStatesTrue,
            addStateFilter: jest.fn(),
            filterByFact: jest.fn(),
            handleFactFilter: jest.fn(),
            clearAllFactFilters: jest.fn(),
            clearAllSelections: jest.fn(),
            setHistory: jest.fn()
        };
    });

    it('should render correctly', () => {
        const wrapper = shallow(
            <DriftToolbar { ...props } />
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    /*it('should render with fact filter chips', () => {
        props.factFilter = 'dog';
        props.activeFactFilters = [ 'cat', 'mouse' ];

        const wrapper = shallow(
            <DriftToolbar { ...props } />
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });*/

    it('should call clearComparisonFilters', () => {
        const wrapper = shallow(
            <DriftToolbar { ...props } />
        );
        wrapper.instance().clearFilters();
        expect(props.clearComparisonFilters).toHaveBeenCalled();
    });

    it('should call clearComparison', () => {
        const wrapper = shallow(
            <DriftToolbar { ...props } />
        );
        wrapper.instance().clearComparison();
        expect(props.clearComparison).toHaveBeenCalled();
    });

    it('should toggle dropdownOpen', () => {
        const wrapper = shallow(
            <DriftToolbar { ...props } />
        );
        wrapper.instance().onToggle();
        expect(wrapper.state('dropdownOpen')).toEqual(true);
    });

    it('should clear state chips', () => {
        const wrapper = shallow(
            <DriftToolbar { ...props } />
        );
        wrapper.instance().clearAllStateChips();

        expect(props.addStateFilter).toHaveBeenCalledWith(
            { filter: 'SAME', display: 'Same', selected: true }
        );
        expect(props.addStateFilter).toHaveBeenCalledWith(
            { filter: 'DIFFERENT', display: 'Different', selected: true }
        );
        expect(props.addStateFilter).toHaveBeenCalledWith(
            { filter: 'INCOMPLETE_DATA', display: 'Incomplete data', selected: true }
        );
    });

    it('should remove single state chip', () => {
        const wrapper = shallow(
            <DriftToolbar { ...props } />
        );
        wrapper.instance().removeChip('State', 'Same');

        expect(props.addStateFilter).toHaveBeenCalledWith(
            { filter: 'SAME', display: 'Same', selected: true }
        );
        expect(props.setHistory).toHaveBeenCalled();
    });

    it('should remove all state chips', () => {
        const wrapper = shallow(
            <DriftToolbar { ...props } />
        );
        wrapper.instance().removeChip('State', '');

        expect(props.addStateFilter).toHaveBeenCalledWith(
            { filter: 'SAME', display: 'Same', selected: true }
        );
        expect(props.addStateFilter).toHaveBeenCalledWith(
            { filter: 'DIFFERENT', display: 'Different', selected: true }
        );
        expect(props.addStateFilter).toHaveBeenCalledWith(
            { filter: 'INCOMPLETE_DATA', display: 'Incomplete data', selected: true }
        );
    });

    it('should remove fact filter chip', async () => {
        const wrapper = shallow(
            <DriftToolbar { ...props } />
        );
        wrapper.instance().removeChip('Fact name', 'blah');

        await expect(props.filterByFact).toHaveBeenCalledWith('');
        expect(props.setHistory).toHaveBeenCalled();
    });

    it('should remove active fact filter chip', () => {
        props.activeFactFilters = [ 'cat', 'mouse' ];
        const wrapper = shallow(
            <DriftToolbar { ...props } />
        );
        wrapper.instance().removeChip('Fact name', 'cat');

        expect(props.handleFactFilter).toHaveBeenCalledWith('cat');
    });

    it('should remove all fact filter chips', () => {
        const wrapper = shallow(
            <DriftToolbar { ...props } />
        );
        wrapper.instance().removeChip('Fact name', '');

        expect(props.clearAllFactFilters).toHaveBeenCalled();
    });

    it.skip('should call clearFilters', () => {
        props.factFilter = 'blah';
        const wrapper = shallow(
            <DriftToolbar { ...props } />
        );
        wrapper.find('.pf-c-button').simulate('click');
        expect(props.clearComparisonFilters).toHaveBeenCalled();
    });
});
