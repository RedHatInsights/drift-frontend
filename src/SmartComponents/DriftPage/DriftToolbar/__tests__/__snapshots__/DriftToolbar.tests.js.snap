// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DriftToolbar should call clearFilters 1`] = `
<Fragment>
  <Toolbar
    className="drift-toolbar"
    clearAllFilters={[Function]}
  >
    <ToolbarContent
      isExpanded={false}
      showClearFiltersButton={false}
    >
      <ForwardRef
        variant="filter-group"
      >
        <ToolbarFilter
          categoryName="Fact name"
          chips={
            Array [
              "blah",
            ]
          }
          deleteChip={[Function]}
          deleteChipGroup={[Function]}
          showToolbarItem={true}
        >
          <SearchBar
            activeFactFilters={Array []}
            factFilter="blah"
            filterByFact={[MockFunction]}
            handleFactFilter={[MockFunction]}
            setHistory={[MockFunction]}
          />
        </ToolbarFilter>
        <ToolbarFilter
          categoryName="State"
          chips={
            Array [
              "Same",
              "Different",
              "Incomplete data",
            ]
          }
          deleteChip={[Function]}
          deleteChipGroup={[Function]}
          showToolbarItem={true}
        >
          <Connect(FilterDropDown)
            setHistory={[MockFunction]}
          />
        </ToolbarFilter>
      </ForwardRef>
      <ForwardRef
        variant="button-group"
      >
        <ToolbarItem>
          <Connect(AddSystemButton)
            loading={false}
          />
        </ToolbarItem>
      </ForwardRef>
      <ForwardRef
        variant="icon-button-group"
      >
        <ToolbarItem>
          <ExportCSVButton
            dropdownItems={
              Array [
                <DropdownItem
                  component="button"
                  data-ouia-component-id="export-to-csv-dropdown-item-comparison"
                  onClick={[Function]}
                >
                  Export to CSV
                </DropdownItem>,
              ]
            }
            isOpen={false}
            onToggle={[Function]}
            ouiaId="export-dropdown-comparison"
          />
        </ToolbarItem>
        <ToolbarItem>
          <ActionKebab
            dropdownItems={
              Array [
                <DropdownItem
                  component="button"
                  data-ouia-component-id="clear-all-comparisons-dropdown-item"
                  onClick={[Function]}
                >
                  Clear all comparisons
                </DropdownItem>,
              ]
            }
            ouiaId="clear-comparison-dropdown"
          />
        </ToolbarItem>
      </ForwardRef>
      <ToolbarItem
        align={
          Object {
            "default": "alignRight",
          }
        }
        variant="pagination"
      >
        <TablePagination
          isCompact={true}
          ouiaId="comparison-pagination-top"
          page={1}
          perPage={50}
          total={30}
          updatePagination={[MockFunction]}
          variant="top"
          widgetId="drift-pagination-top"
        />
      </ToolbarItem>
    </ToolbarContent>
  </Toolbar>
</Fragment>
`;

exports[`DriftToolbar should render correctly 1`] = `
<Fragment>
  <Toolbar
    className="drift-toolbar"
    clearAllFilters={[Function]}
    clearFiltersButtonText="Reset filters"
  >
    <ToolbarContent
      isExpanded={false}
      showClearFiltersButton={false}
    >
      <ForwardRef
        variant="filter-group"
      >
        <ToolbarFilter
          categoryName="Fact name"
          chips={Array []}
          deleteChip={[Function]}
          deleteChipGroup={[Function]}
          showToolbarItem={true}
        >
          <SearchBar
            activeFactFilters={Array []}
            factFilter=""
            filterByFact={[MockFunction]}
            handleFactFilter={[MockFunction]}
            setHistory={[MockFunction]}
          />
        </ToolbarFilter>
        <ToolbarFilter
          categoryName="State"
          chips={
            Array [
              "Same",
              "Different",
              "Incomplete data",
            ]
          }
          deleteChip={[Function]}
          deleteChipGroup={[Function]}
          showToolbarItem={true}
        >
          <Connect(FilterDropDown)
            setHistory={[MockFunction]}
          />
        </ToolbarFilter>
      </ForwardRef>
      <ForwardRef
        variant="button-group"
      >
        <ToolbarItem>
          <Connect(AddSystemButton)
            loading={false}
          />
        </ToolbarItem>
      </ForwardRef>
      <ForwardRef
        variant="icon-button-group"
      >
        <ToolbarItem>
          <ExportCSVButton
            dropdownItems={
              Array [
                <DropdownItem
                  component="button"
                  data-ouia-component-id="export-to-csv-dropdown-item-comparison"
                  onClick={[Function]}
                >
                  Export to CSV
                </DropdownItem>,
                <DropdownItem
                  component="button"
                  data-ouia-component-id="export-to-json-dropdown-item-comparison"
                  onClick={[Function]}
                >
                  Export to JSON
                </DropdownItem>,
              ]
            }
            isOpen={false}
            onToggle={[Function]}
            ouiaId="export-dropdown-comparison"
          />
        </ToolbarItem>
        <ToolbarItem>
          <ActionKebab
            dropdownItems={
              Array [
                <DropdownItem
                  component="button"
                  data-ouia-component-id="clear-all-comparisons-dropdown-item"
                  onClick={[Function]}
                >
                  Clear all comparisons
                </DropdownItem>,
              ]
            }
            ouiaId="clear-comparison-dropdown"
          />
        </ToolbarItem>
      </ForwardRef>
      <ToolbarItem
        align={
          Object {
            "default": "alignRight",
          }
        }
        variant="pagination"
      >
        <TablePagination
          isCompact={true}
          ouiaId="comparison-pagination-top"
          page={1}
          perPage={50}
          total={30}
          updatePagination={[MockFunction]}
          variant="top"
          widgetId="drift-pagination-top"
        />
      </ToolbarItem>
    </ToolbarContent>
  </Toolbar>
</Fragment>
`;

exports[`DriftToolbar should render with fact filter chips 1`] = `
<Fragment>
  <Toolbar
    className="drift-toolbar"
    clearAllFilters={[Function]}
    clearFiltersButtonText="Reset filters"
  >
    <ToolbarContent
      isExpanded={false}
      showClearFiltersButton={false}
    >
      <ForwardRef
        variant="filter-group"
      >
        <ToolbarFilter
          categoryName="Fact name"
          chips={
            Array [
              "cat",
              "mouse",
              "dog",
            ]
          }
          deleteChip={[Function]}
          deleteChipGroup={[Function]}
          showToolbarItem={true}
        >
          <SearchBar
            activeFactFilters={
              Array [
                "cat",
                "mouse",
              ]
            }
            factFilter="dog"
            filterByFact={[MockFunction]}
            handleFactFilter={[MockFunction]}
            setHistory={[MockFunction]}
          />
        </ToolbarFilter>
        <ToolbarFilter
          categoryName="State"
          chips={
            Array [
              "Same",
              "Different",
              "Incomplete data",
            ]
          }
          deleteChip={[Function]}
          deleteChipGroup={[Function]}
          showToolbarItem={true}
        >
          <Connect(FilterDropDown)
            setHistory={[MockFunction]}
          />
        </ToolbarFilter>
      </ForwardRef>
      <ForwardRef
        variant="button-group"
      >
        <ToolbarItem>
          <Connect(AddSystemButton)
            loading={false}
          />
        </ToolbarItem>
      </ForwardRef>
      <ForwardRef
        variant="icon-button-group"
      >
        <ToolbarItem>
          <ExportCSVButton
            dropdownItems={
              Array [
                <DropdownItem
                  component="button"
                  data-ouia-component-id="export-to-csv-dropdown-item-comparison"
                  onClick={[Function]}
                >
                  Export to CSV
                </DropdownItem>,
                <DropdownItem
                  component="button"
                  data-ouia-component-id="export-to-json-dropdown-item-comparison"
                  onClick={[Function]}
                >
                  Export to JSON
                </DropdownItem>,
              ]
            }
            isOpen={false}
            onToggle={[Function]}
            ouiaId="export-dropdown-comparison"
          />
        </ToolbarItem>
        <ToolbarItem>
          <ActionKebab
            dropdownItems={
              Array [
                <DropdownItem
                  component="button"
                  data-ouia-component-id="clear-all-comparisons-dropdown-item"
                  onClick={[Function]}
                >
                  Clear all comparisons
                </DropdownItem>,
              ]
            }
            ouiaId="clear-comparison-dropdown"
          />
        </ToolbarItem>
      </ForwardRef>
      <ToolbarItem
        align={
          Object {
            "default": "alignRight",
          }
        }
        variant="pagination"
      >
        <TablePagination
          isCompact={true}
          ouiaId="comparison-pagination-top"
          page={1}
          perPage={50}
          total={30}
          updatePagination={[MockFunction]}
          variant="top"
          widgetId="drift-pagination-top"
        />
      </ToolbarItem>
    </ToolbarContent>
  </Toolbar>
</Fragment>
`;
